File: examples/todolist/core/change/additem/add_item.go
package change

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"

	"github.com/err0r500/fairway"
	"github.com/err0r500/fairway/examples/todolist/core/change"
	"github.com/err0r500/fairway/examples/todolist/core/event"
)

func init() {
	change.ChangeRegistry.RegisterCommand("POST /api/lists/{listId}/items/{itemId}", httpHandler)
}

var itemAlreadyExistsErr = errors.New("item already exists")

type reqBody struct {
	Text string `json:"text" validate:"required"`
}

func httpHandler(runner fairway.CommandRunner) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req reqBody
		if !fairway.JsonParse(w, r, &req) {
			return
		}

		if err := runner.RunPure(r.Context(), command{
			listId: r.PathValue("listId"),
			itemId: r.PathValue("itemId"),
			text:   req.Text,
		}); err != nil {
			switch err {
			case itemAlreadyExistsErr:
				w.WriteHeader(http.StatusConflict)

			default:
				w.WriteHeader(http.StatusInternalServerError)
				json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
			}
			return
		}

		w.WriteHeader(http.StatusCreated)
	}
}

type command struct {
	listId string
	itemId string
	text   string
}

func (cmd command) Run(ctx context.Context, ev fairway.EventReadAppender) error {
	itemAlreadyExists := false

	if err := ev.ReadEvents(ctx,
		fairway.QueryItems(
			fairway.NewQueryItem().
				Types(event.ItemAdded{}).
				Tags(event.ItemTagPrefix(cmd.itemId)),
		),
		func(te fairway.TaggedEvent) bool {
			switch te.(type) {
			case event.ItemAdded:
				itemAlreadyExists = true
				return false
			default:
				return true
			}
		}); err != nil {
		return err
	}

	if itemAlreadyExists {
		return itemAlreadyExistsErr
	}

	return ev.AppendEvents(ctx,
		fairway.TaggedEvent(
			event.ItemAdded{ListId: cmd.listId, ItemId: cmd.itemId, Text: cmd.text},
		))
}

File: fairway_ex/lib/examples/todolist/change/add_item.ex
defmodule Todolist.Change.AddItem do
  @moduledoc """
  Pure command: add an item to an existing todo list.

  Checks whether the item already exists (by `item_id` + `list_id`) via DCB.
  Returns `{:error, :already_exists}` if the item was already added.
  """

  alias Fairway.{Query, QueryItem, ReadAppender, Event}
  alias Todolist.Event.ItemCreated

  def run(ra, %{list_id: list_id, item_id: item_id, text: text}) do
    query = Query.new([
      QueryItem.new(
        types: [ItemCreated],
        tags: ["list_id:#{list_id}", "item_id:#{item_id}"]
      )
    ])

    {ra, exists} =
      ReadAppender.fold_events(ra, query, false, fn
        %{data: %ItemCreated{}}, _ -> {false, true}
        _, acc -> {true, acc}
      end)

    if exists do
      {:error, :already_exists}
    else
      ReadAppender.append_events(ra, [
        Event.new(%ItemCreated{list_id: list_id, item_id: item_id, text: text})
      ])
    end
  end
end
